### PR TITLE
add workflow activation and scheduling lock TTL

### PR DESCRIPTION
**Pull request by garden-of-delete**

### CODE REVIEW

File: service/scheduler.go [hunk 1]
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
...
 }
 
 func (s *Scheduler) Start() {
 	fmt.Println("Scheduler Started")
 	tick := time.Tick(s.Interval)
-	for range tick {
-		fmt.Println("tick")
-		s.scheduleWorkflows(database.GetInstance())
-		s.activateWorkflows(database.GetInstance())
+	lockTtlTick := time.Tick(time.Minute * 5) // TODO: ok to hardcode to 5 mins?
```
Comment (garden-of-delete) [inline service/scheduler.go:68]: Hardcoded to check for and delete locks older than LockTTL every 5 minutes. Could also do it every second, or parameterize the interval in the config.

File: cmd/scheduler.go [hunk 2]
```diff
@@ -82,4 +85,11 @@ func init() {
 		"api key to orchard service",
 	)
 	viper.BindPFlag("scheduler.orchard.apiKey", schedulerCmd.Flags().Lookup("orchard"))
+
+	schedulerCmd.Flags().Duration(
+		"lockTTL",
+		time.Hour,
```
Comment (garden-of-delete) [inline cmd/scheduler.go:91]: is one hour a good default duration for the LockTTL?

Comment (schowsf) [reply cmd/scheduler.go:91]: I think more flexible to have in units of time.Minute

File: service/scheduler.go (hunk 1 above)
Comment (garden-of-delete) [reply service/scheduler.go]: Moved the default lockTTL up to 15 mins per our convo

File: cmd/scheduler.go (hunk 2 above)
Comment (garden-of-delete) [reply cmd/scheduler.go]: Agreed, done.

File: service/scheduler.go [hunk 3]
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
...
+			s.scheduleWorkflows(database.GetInstance())
+			s.activateWorkflows(database.GetInstance())
+		case <-lockTtlTick:
+			fmt.Println("lockTtlTick")
+			s.deleteOldLocks(database.GetInstance())
+		}
 	}
 }
 
+func (s *Scheduler) deleteOldLocks(db *gorm.DB) {
```
Comment (garden-of-delete) [inline service/scheduler.go:82]: handles the case where the pod crashes before the lock can be released in the lockAndCreate / lockAndActivate methods

File: service/scheduler.go [hunk 4]
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
...
 }
 
 func (s *Scheduler) Start() {
 	fmt.Println("Scheduler Started")
 	tick := time.Tick(s.Interval)
-	for range tick {
-		fmt.Println("tick")
-		s.scheduleWorkflows(database.GetInstance())
-		s.activateWorkflows(database.GetInstance())
+	lockTtlTick := time.Tick(time.Minute)
```
[the remark this answers is not in the document]
Comment (garden-of-delete) [reply service/scheduler.go:68]: In this implementation, `lockTtlTick` is the tick that defines the frequency that the scheduler checks for expired locks. LockTTL is a separate value that defines when a lock is considered expires.

File: service/scheduler.go [hunk 5]
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
...
+	for {
+		select {
+		case <-tick:
+			fmt.Println("tick")
+			s.scheduleWorkflows(database.GetInstance())
+			s.activateWorkflows(database.GetInstance())
+		case <-lockTtlTick:
```
[the remark this answers is not in the document]
Comment (garden-of-delete) [reply service/scheduler.go:75]: I considered that option. Just wasn't sure if we wanted to be checking for and removing expired locks every time the scheduler `lockAndCreate`s. I can move it and remove the separate ticker for that check.

**Comment by garden-of-delete** [comment]
Moved the "check for and delete expired locks" logic into the `lockAndCreate` method. Also refactored `lockTTL` to `lockTimeout` and `deleteOldLocks` to `deleteExpiredLocks`.

**Pull request merged**